src/backend/d3d12/ResourceUploader.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ResourceUploader.h"
+
+#include "D3D12Backend.h"
+
+namespace backend {
+namespace d3d12 {
+
+    ResourceUploader::ResourceUploader(Device* device) : device(device) {
+        ASSERT_SUCCESS(device->GetD3D12Device()->CreateFence(uploadSerial, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence)));
+    }
+
+    void ResourceUploader::UploadResource(ComPtr<ID3D12Resource> resource, uint32_t start, uint32_t count, const uint32_t* data) {
+        D3D12_RESOURCE_DESC resourceDescriptor = resource->GetDesc();
```

Comment:
This function is using CopyBufferRegion, so it requires both resources to be buffers and you can easily make your own resource descriptor.

The comment was about this function creating a full-sized buffer on the upload heap: if ```resource``` is several megabytes and we want to update a couple bytes, it would be more efficient to just create a ```uploadResource``` of a couple bytes. Instead right now, the function does an allocation of a several megabytes and uses just a couple byte of it.

Basically uploadResource needs to have a size of ```count``` and CopyBufferRegion's fourth argument becomes 0. (and the map + memcpy logic needs to be updated too)